examples/images_contours_and_fields/irregulardatagrid.py

```diff
@@ -0,0 +1,105 @@
+"""
+=======================================
+Contour plot of irregularly spaced data
+=======================================
+
+Comparison of a contour plot of irregularly spaced data interpolated
+on a regular grid versus a tricontour plot for an unstructured triangular grid.
+
+Since :meth:`~.axes.Axes.contour` and :meth:`~.axes.Axes.contourf` expect the
+data to live on a regular grid, plotting a contour plot of irregularly spaced
+data requires different methods. The two options are:
+
+* Interpolate the data to a regular grid first. This can be done with on-borad
+  means, e.g. via `~.tri.LinearTriInterpolator` or using external functionality
+  e.g. via `scipy.interpolate.griddata`. The latter is more versatile
+  as it allows for different kinds of interpolation (e.g. cubic). Then plot the
+  interpolated data with the usual :meth:`~.axes.Axes.contour`.
+* Directly use :meth:`~.axes.Axes.tricontour` or :meth:`~.axes.Axes.tricontourf`
+  which will perform a triangulation internally.
+
+This example shows both methods in action.
+"""
+
+import matplotlib.pyplot as plt
+import matplotlib.tri as tri
+import numpy as np
+
+np.random.seed(19680801)
+npts = 200
+ngridx = 100
+ngridy = 200
+x = np.random.uniform(-2, 2, npts)
+y = np.random.uniform(-2, 2, npts)
+z = x * np.exp(-x**2 - y**2)
+
+fig, (ax1, ax2) = plt.subplots(nrows=2)
+
+# -----------------------
+# Interpolation on a grid
+# -----------------------
+# A contour plot of irregularly spaced data coordinates
+# via interpolation on a grid.
+
+# Create grid values first.
+xi = np.linspace(-2.1, 2.1, ngridx)
+yi = np.linspace(-2.1, 2.1, ngridy)
+
+# Perform linear interpolation of the data (x,y)
+# on a grid defined by (xi,yi)
+triang = tri.Triangulation(x, y)
+interpolator = tri.LinearTriInterpolator(triang, z)
```

Comment:
OK, Hmmmm.  I'm not 100% convinced that we should suggest people use the in-house triangular and interpolator.  In the same vein that `mlab` is going away, my guess is that `tri` will go away.  I'd rather we tell people the "best practice" and I'm not sure our sparsely documented in-house triangulation routine is that.  Happy to be corrected, but I'd lean towards the scipy reference.

Reply:
@jklymak I assure you that `tri` is not going to "go away". What on earth possessed you to assert such a guess?

I am 100% convinced that we should suggest people use the in-house triangulator and interpolator, and unlike yourself I am an expert in the area. Consider yourself corrected, and in future kindly refrain from expressing derogatory opinions on functionality you have no knowledge of.

Reply:
I don't think what @jklymak said was particularly derogatory...

Is there a reason Matplotlib runs a custom triangulator, and doesn't use an upstream one? And if ours is better than what is available in either numpy or scipy, should we try and push the mpl triangulation code upstream? It feels to me like that's where good triangulation code should live.

Reply:
I don't think this is about triangulation. It's about the interpolators from `tri.triinterpolate`.  
I cannot judge on whether they are needed or not, but here are the timings from the example:
```
* LinearTriInterpolator   0.0444 s
* scipy griddata          0.0728 s
* matplotlib tricontour   0.0301 s
```

Reply:
Just out of curiosity, not because I think performance is particularly the point:  On my machine, `griddata` is definitely pokey for small data sets.  Larger data sets `griddata` starts being faster than the matplotlib implementation.  Basically `griddata` has a 0.16 s overhead for some reason.

The less-convenient `LinearNDInterpolator` is faster than both for all data set sizes (on my machine). 

Obviously speed isn't the only metric of a good interpolation scheme, so I'm *definitely* not claiming the script scheme is better.  I'm just not sure that our internal triangulation/interpolation method should be heavily advertised compared to the more standard `scipy` one, unless there is a compelling reason ours is superior.  And if ours is superior, why aren't we sharing the love upstream?  

```

Npts =    200
tri                  0.004273891448974609
griddata             0.16399312019348145
LinearNDInterpolator 0.002990245819091797

Npts =   2000
tri                  0.029674053192138672
griddata             0.16156911849975586
LinearNDInterpolator 0.016344070434570312

Npts =  20000
tri                  0.3896908760070801
griddata             0.3552122116088867
LinearNDInterpolator 0.1670820713043213

Npts = 200000
tri                  5.315340995788574
griddata             2.238363265991211
LinearNDInterpolator 2.089240074157715

```

```python


import matplotlib.pyplot as plt
import matplotlib.tri as tri
import numpy as np
import time
import scipy.interpolate as sinterp

np.random.seed(19680801)
npts = 200000
ngridx = 100
ngridy = 200
x = np.random.uniform(-2, 2, npts)
y = np.random.uniform(-2, 2, npts)
z = x * np.exp(-x**2 - y**2)

xi = np.linspace(-2.1, 2.1, ngridx)
yi = np.linspace(-2.1, 2.1, ngridy)
print('Npts = %6d'%npts)

time0 = time.time()
triang = tri.Triangulation(x, y)
interpolator = tri.LinearTriInterpolator(triang, z)
Xi, Yi = np.meshgrid(xi, yi)
zi = interpolator(Xi, Yi)
time1 = time.time()
print('tri                 ', time1-time0)

time0 = time.time()
zi = sinterp.griddata((x, y), z, (xi[None,:], yi[:,None]), method='linear')
time1 = time.time()
print('griddata            ', time1-time0)

time0 = time.time()
interp = sinterp.LinearNDInterpolator((x, y), z)
zi = interp(Xi, Yi)
time1 = time.time()
print('LinearNDInterpolator', time1-time0)
```

Reply:
There is no 'upstream', no de facto standard library for representing and manipulating triangular grids in python. If there was, we should certainly use it. Until then, we are one of the gathering places for such functionality. It started as just rendering triangular grids, which is clearly within our remit, and when interpolation and mesh refinement were implemented they were added to our code as the best (least bad?) place to put them as our code is actively maintained and will be around for a long time.

Numpy isn't the place for such functionality, scipy would be more appropriate but I don't think it is as suitable as a specific tri grids library (named `tripy` perhaps?). Every so often there is an interest in such as library, but after lots of chat my summary is that there are plenty of people who want to **use** such a library, nobody wants to **write** it. From my point of view, even if a `tripy` existed the rendering of tri grids, which is what I am mostly interested it, would still reside within mpl.

Without delving too deeply into the workings of our tri code, there are three use cases for creating and using a triangulation object:
1. User specifies the 2D points and requests a Delaunay triangulation is calculated for them.
2. The same as 1, and then the user masks out some of the unwanted triangles.
3. User specifies both the 2D points and the triangulation.

scipy.griddata can only deal with use case 1, our code deals with all three. Unfortunately use case 1 is the one that most examples refer to as it is the entry point to triangular grids for those not used to them. But nearly all serious tri grid work uses the other 2 use cases.

Hence we should not recommend using scipy.griddata, it simply doesn't include the functionality that our code requires and has. But unfortunately we can't just ignore it and not mention it in such examples; we need to keep it as there are lots of historical web pages mentioning griddata, and we need to catch the attention of people looking for it so that we can point out our alternatives. Keeping it in this example but just within a comment is an approach that I like.

Unfortunately griddata have been around for years and people are familiar with it, so anyone who asks about tri grids receives a well-intended but sub-optimal answer to interpolate to a regular grid. It is simply unnecessary. Everyone should use `tricontour` for optimal results: the boundaries and masked areas are correct, it has higher resolution where the grid has higher resolution. If higher resolution is required, there is an excellent `TriRefiner` that creates a new triangulation, keeping sensible boundaries, etc and can avoid badly-formed triangles (e.g. long and thin); it is featured in the `tricontour_smooth_delaunay` and `tricontour_smooth_user` examples.